transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -979,6 +982,26 @@ final void setAddPending() {
         assert updated; // This should always be true as it MUST be called before setAddComplete() or setRemoved().
     }
 
+    final void callHandlerAdded() throws Exception {
+        // We must call setAddComplete before calling handlerAdded. Otherwise if the handlerAdded method generates
+        // any pipeline events ctx.handler() will miss them because the state will not allow it.
+        if (setAddComplete()) {
+            handler().handlerAdded(this);
+        }
+    }
+
+    final void callHandlerRemoved() throws Exception {
+        try {
+            // Only call handlerRemoved(...) if we called handlerAdded(...) before.
+            if (handlerState == ADD_COMPLETE) {
+                handler().handlerRemoved(this);
+            }
+        } finally {
+            // Mark the handler as removed in any case.
+            setRemoved();
```

Comment:
Not a part of this change, but it's surprising to me that we call `setRemoved()` _after_ we call `handlerRemoved`. For example, if the handler decides to call `ctx.isRemoved()` in it's `handlerRemoved` method it will report `false` which seems inconsistent.

Reply:
@bryce-anderson let me open another issue to investigate this.